src/modules/roc_fec/writer.cpp

```diff
@@ -68,51 +69,85 @@ void Writer::write(const packet::PacketPtr& pp) {
         } while (source_ == pp->rtp()->source);
     }
 
+    if (cur_packet_ == 0) {
+        if (!begin_block_()) {
+            return;
+        }
+    }
+
+    write_source_packet_(pp);
+
+    cur_packet_++;
+
+    if (cur_packet_ == cur_sblen_) {
+        end_block_();
+        next_block_();
+    }
+}
+
+void Writer::resize(size_t sblen) {
+    roc_log(LogDebug, "fec writer: update sblen, cur=%lu next=%lu",
+            (unsigned long)cur_sblen_, (unsigned long)sblen);
+
+    next_sblen_ = sblen;
+
+    if (cur_packet_ == 0) {
+        cur_sblen_ = sblen;
+    }
+}
+
+void Writer::write_source_packet_(const packet::PacketPtr& pp) {
+    encoder_.set(cur_packet_, pp->fec()->payload);
+
     pp->add_flags(packet::Packet::FlagComposed);
     fill_packet_fec_fields_(pp, (packet::seqnum_t)cur_packet_);
 
     if (!source_composer_.compose(*pp)) {
         roc_panic("fec writer: can't compose packet");
     }
+
     writer_.write(pp);
+}
 
-    encoder_.set(cur_packet_, pp->fec()->payload);
-    cur_packet_++;
+bool Writer::begin_block_() {
+    if (encoder_.begin(cur_sblen_, cur_rblen_)) {
+        return true;
+    }
+
+    roc_log(LogError, "fec writer: can't begin encoding");
+    return false;
+}
 
-    if (cur_packet_ == n_source_packets_) {
-        handle_next_block_();
+void Writer::end_block_() {
+    make_repair_packets_();
 
-        cur_block_repair_sn_ += n_repair_packets_;
-        cur_sbn_++;
+    encode_repair_packets_();
+    encoder_.fill();
 
-        cur_packet_ = 0;
-    }
+    write_repair_packets_();
+
+    encoder_.end();
 }
 
-void Writer::handle_next_block_() {
+void Writer::next_block_() {
+    cur_block_repair_sn_ += cur_rblen_;
+    cur_sbn_++;
+
+    cur_packet_ = 0;
+    cur_sblen_ = next_sblen_;
+
     roc_log(LogTrace, "fec writer: next block: sbn=%lu", (unsigned long)cur_sbn_);
```

Comment:
Please also add sblen and rblen to the log message.

Reply:
Done